plugins/crate-copy-azure/src/main/java/io/crate/copy/azure/AzureFileInput.java

```diff
@@ -0,0 +1,147 @@
+    private final URI uri;
+    private final GlobPredicate uriPredicate;
+    private final String preGlobPath;
+    private final Operator operator;
+    private final SharedAsyncExecutor sharedAsyncExecutor;
+
+    /**
+     * @param uri is in user provided format (azblob://path/to/dir)
+     */
+    public AzureFileInput(SharedAsyncExecutor sharedAsyncExecutor, URI uri, Settings settings) {
+        this.config = AzureBlobStorageSettings.openDALConfig(settings, CopyStatementSettings.commonCopyFromSettings);
+        this.sharedAsyncExecutor = sharedAsyncExecutor;
+        this.operator = AsyncOperator.of(NAME, config, sharedAsyncExecutor.asyncExecutor()).blocking();
+        // Pre-glob path operates with user-provided URI as GLOB pattern reflects user facing COPY FROM uri format.
+        this.preGlobPath = toPreGlobPath(uri);
+
+        String resourcePath = resourcePath(uri);
+        this.uri = URI.create(resourcePath);
+        // Glob predicate operates with normalized resource URI to reflect OpenDAL List API response format.
+        // List API returns entries without leading backslash.
+        this.uriPredicate = new GlobPredicate(resourcePath.substring(1));
+    }
+
+    /**
+     * @return List<URI> in Azure compatible format.
+     */
+    @Override
+    public List<URI> expandUri() throws IOException {
+        var operator = operator(); // Only for tests, otherwise could use this.operator
+        if (isGlobbed() == false) {
+            return List.of(uri);
+        }
+        List<URI> uris = new ArrayList<>();
+        List<Entry> entries = operator.list(preGlobPath);
+        for (Entry entry : entries) {
+            var path = entry.getPath();
+            if (uriPredicate.test(path)) {
+                uris.add(URI.create(path));
+            }
+        }
+        return uris;
+    }
+
+    @VisibleForTesting
+    Operator operator() {
+        return AsyncOperator.of(NAME, config, sharedAsyncExecutor.asyncExecutor()).blocking();
+    }
+
+    /**
+     * @param uri is resource path without "azblob" schema.
+     */
+    @Override
+    public InputStream getStream(URI uri) throws IOException {
+        return operator.createInputStream(uri.toString());
+    }
+
+    @Override
+    public boolean isGlobbed() {
+        return preGlobPath != null;
+    }
+
+    @Override
+    public URI uri() {
+        return uri;
+    }
+
+    @Override
+    public boolean sharedStorageDefault() {
+        return true;
+    }
+
+    @Override
+    public void close() {
+        assert operator != null : "Operator must be created before FileInput is closed";
+        operator.close();
+    }
+
+    /**
+     * @return pre-glob path in Azure compatible format.
+     */
+    @Nullable
+    public static String toPreGlobPath(URI uri) {
+        Matcher hasGlobMatcher = HAS_GLOBS_PATTERN.matcher(uri.toString());
+        if (hasGlobMatcher.matches()) {
+            return resourcePath(URI.create(hasGlobMatcher.group(1)));
```

Comment:
I guess the `URI.create()` roundtrip is purely for URI string parsing validation. If already validate it wouldn't make any sense.
```suggestion
            return URI.create(hasGlobMatcher.group(1)).toString();
```